File: platform/platform-impl/src/com/intellij/ide/actions/WhatsNewAction.java
// Copyright 2000-2026 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.ide.actions;

import com.intellij.icons.AllIcons;
import com.intellij.ide.BrowserUtil;
import com.intellij.ide.IdeBundle;
import com.intellij.notification.NotificationAction;
import com.intellij.notification.NotificationType;
import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.application.ApplicationInfo;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ApplicationNamesInfo;
import com.intellij.openapi.application.IdeUrlTrackingParametersProvider;
import com.intellij.openapi.fileEditor.impl.HTMLEditorProvider;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.updateSettings.impl.UpdateCheckerFacade;
import com.intellij.platform.ide.customization.ExternalProductResourceUrls;
import com.intellij.util.ui.accessibility.ScreenReader;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.event.ActionEvent;

@ApiStatus.Internal
public final class WhatsNewAction extends AnAction implements DumbAware {
  @Override
  public void update(@NotNull AnActionEvent e) {
    var available = ExternalProductResourceUrls.getInstance().getWhatIsNewPageUrl() != null;
    e.getPresentation().setEnabledAndVisible(available);
    if (available) {
      e.getPresentation().setText(IdeBundle.messagePointer("whats.new.action.custom.text", ApplicationNamesInfo.getInstance().getFullProductName()));
      e.getPresentation().setDescription(IdeBundle.messagePointer("whats.new.action.custom.description", ApplicationNamesInfo.getInstance().getFullProductName()));
    }
  }

  @Override
  public @NotNull ActionUpdateThread getActionUpdateThread() {
    return ActionUpdateThread.BGT;
  }

  @Override
  public void actionPerformed(@NotNull AnActionEvent e) {
    var whatsNewUrl = ExternalProductResourceUrls.getInstance().getWhatIsNewPageUrl();
    if (whatsNewUrl == null) throw new IllegalStateException();
    var url = whatsNewUrl.toExternalForm();

    if (ApplicationManager.getApplication().isInternal() && (e.getModifiers() & ActionEvent.SHIFT_MASK) != 0) {
      var title = IdeBundle.message("whats.new.action.custom.text", ApplicationNamesInfo.getInstance().getFullProductName());
      var prompt = IdeBundle.message("browser.url.popup");
      url = Messages.showInputDialog(e.getProject(), prompt, title, null, url, null);
      if (url == null) return;
    }

    var project = e.getProject();
    if (project != null && openEmbeddedWhatsNewPage(project, url, false, null)) {
      return;
    }

    BrowserUtil.browse(IdeUrlTrackingParametersProvider.getInstance().augmentUrl(url));
  }

  @ApiStatus.Internal
  public static void openWhatsNewPage(@NotNull Project project, @NotNull String url, boolean onUpgrade) {
    if (!ScreenReader.isActive() && openEmbeddedWhatsNewPage(project, url, onUpgrade, null)) {
      return;
    }

    var name = ApplicationNamesInfo.getInstance().getFullProductName();
    var version = ApplicationInfo.getInstance().getShortVersion();
    String notificationText =
      IdeBundle.message(ScreenReader.isActive() ? "whats.new.notification.text.regular.language" : "whats.new.notification.text", name, version);
    UpdateCheckerFacade.getInstance().getNotificationGroupForIdeUpdateResults()
      .createNotification(notificationText, NotificationType.INFORMATION)
      .setIcon(AllIcons.Nodes.PpWeb)
      .setDisplayId("ide.whats.new")
      .addAction(NotificationAction.createSimpleExpiring(IdeBundle.message("whats.new.notification.action"), () -> BrowserUtil.browse(url)))
      .notify(project);
  }

  @ApiStatus.Internal
  @SuppressWarnings("unused")
  public static void openWhatsNewPage(@NotNull Project project,
                                      @NotNull String url,
                                      boolean includePlatformData,
                                      @Nullable HTMLEditorProvider.JsQueryHandler queryHandler) {
    if (!openEmbeddedWhatsNewPage(project, url, includePlatformData, queryHandler)) {
      throw new IllegalStateException("Embedded What's New page is not available");
    }
  }

  private static boolean openEmbeddedWhatsNewPage(@NotNull Project project,
                                                  @NotNull String url,
                                                  boolean includePlatformData,
                                                  @Nullable HTMLEditorProvider.JsQueryHandler queryHandler) {
    var provider = WhatsNewPageProvider.EP_NAME.findFirstSafe(WhatsNewPageProvider::isAvailable);
    if (provider == null) return false;
    provider.openWhatsNewPage(project, url, includePlatformData, queryHandler);
    return true;
  }
}

File: platform/whatsNew/src/com/intellij/platform/whatsNew/WhatsNewAction.kt
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.platform.whatsNew

import com.intellij.ide.DataManager
import com.intellij.ide.IdeBundle
import com.intellij.ide.actions.WhatsNewAction
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.application.ApplicationNamesInfo
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.diagnostic.runAndLogException
import com.intellij.openapi.progress.currentThreadCoroutineScope
import com.intellij.openapi.project.Project
import com.intellij.platform.whatsNew.reaction.FUSReactionChecker
import kotlinx.coroutines.launch
import org.jetbrains.concurrency.await

internal class WhatsNewAction : AnAction(), com.intellij.openapi.project.DumbAware {
  companion object {
    internal const val PLACE = "WhatsNew"
  }

  private val REACTIONS_STATE = "whatsnew.reactions.state"
  private val reactionChecker = FUSReactionChecker(REACTIONS_STATE)

  suspend fun openWhatsNew(project: Project) {
    LOG.info("Open What's New page requested.")
    val dataContext = LOG.runAndLogException { DataManager.getInstance().dataContextFromFocusAsync.await() }
    openWhatsNewPage(project, dataContext)
  }

  private suspend fun openWhatsNewPage(project: Project, dataContext: DataContext?, triggeredByUser: Boolean = false) {
    val content = WhatsNewContent.getWhatsNewContent()
    if (content != null && content.isAvailable()) {
      content.show(project, dataContext, triggeredByUser, reactionChecker)
    }
    else {
      LOG.warn("EapWhatsNew: can't be shown. Content not available")
    }
  }

  override fun getActionUpdateThread(): ActionUpdateThread {
    return ActionUpdateThread.BGT
  }

  override fun update(e: AnActionEvent) {
    e.presentation.isEnabledAndVisible = service<WhatsNewStatus>().isContentAvailable
    e.presentation.setText(IdeBundle.messagePointer("whats.new.action.custom.text", ApplicationNamesInfo.getInstance().fullProductName))
    e.presentation.setDescription(IdeBundle.messagePointer("whats.new.action.custom.description", ApplicationNamesInfo.getInstance().fullProductName))
  }

  override fun actionPerformed(e: AnActionEvent) {
    val project = e.project
    if (project != null) {
      currentThreadCoroutineScope().launch {
        openWhatsNewPage(project, e.dataContext, true)
      }
    } else {
      LOG.warn("Cannot open what's new action page because project is null")
    }
  }
}

private val LOG = logger<WhatsNewAction>()
